frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts

```diff
@@ -71,7 +71,7 @@ export const getReduxData = (immutableData, resource: WatchK8sResource) => {
 };
 
 export const getIDAndDispatch: GetIDAndDispatch<SDKStoreState> = (resource, k8sModel, cluster) => {
-  if (!k8sModel || !resource) {
+  if (!k8sModel || !resource || (!resource.isList && !resource.name)) {
```

Comment:
I think we should look at updating the type too, no? If this is not a valid state, we should make TS step in the way and say the call is bad. I think in your case however, the dynamic nature of how you get `name`, TS might not help out here.

In principle, however, I think we should update our types. Thoughts?

Reply:
So `isList` and `name` are optional fields and its reflected in the `WatchK8sResource` as well. Can you please explain which type you are talking about and how should we update it?

Reply:
There is no relationship between the two fields (`isList` & `name`) in Typescript but... you're stating here that `if I am not isList and I do not have a name, that's a bad state`. We should try to showcase that in the Types so we can properly establish typings to match logic and get away from pure runtime checks when we have TypeScript.

### But, as I was testing it out, to no surprise to anyone I'm sure, OpenShift Console has massively poor types and this following solution breaks all kinds of code from a TS point of view. I'll log a bug to address this and you can ignore this comment.

But for your own learning experience, this is what I was thinking...

---

`WatchK8sResource` got a facelift in [openshift/dynamic-plugin-sdk](https://github.com/openshift/dynamic-plugin-sdk/blob/main/packages/lib-utils/src/k8s/hooks/watch-resource-types.ts#L25-L39) to make the API more idiomatic (ie optional everything is a poor type). We could follow that logic and expand it to `!isList` MUST have `name`, imo.

Something like:
```
export type WatchK8sResource = EitherNotBoth<
  { kind: K8sResourceKindReference },
  { groupVersionKind: K8sGroupVersionKind }
> &
  EitherNotBoth<{ isList: true }, { name: string }> &
  Partial<{
    namespace: string;
    selector: Selector;
    namespaced: boolean;
    limit: number;
    fieldSelector: string;
    optional: boolean;
    partialMetadata: boolean;
    cluster: string;
  }>;
```
`EitherNotBoth` Generic type comes [from here](https://github.com/openshift/dynamic-plugin-sdk/blob/main/packages/common/src/types/common.ts#L27-L90).